src/v/storage/tests/batch_consumer_utils_test.cc

```diff
@@ -0,0 +1,200 @@
+    }
+
+    void
+    skip_batch_start(model::record_batch_header hdr, size_t, size_t) override {
+        if (expect_close) {
+            BOOST_FAIL("Unexpected skip_batch_start");
+        }
+        skipped->push_back(hdr);
+    }
+
+    void consume_records(iobuf&&) override {
+        if (expect_close) {
+            BOOST_FAIL("Unexpected consume_records");
+        }
+    }
+
+    ss::future<stop_parser> consume_batch_end() override {
+        if (expect_close) {
+            BOOST_FAIL("Unexpected consume_batch_end");
+        }
+        if (consumed->size() < size_limit) {
+            co_return stop_parser::no;
+        }
+        co_return stop_parser::yes;
+    }
+
+    void print(std::ostream& o) const override { o << "test_consumer"; }
+
+    size_t side{0};
+    size_t size_limit{0};
+    std::deque<model::record_batch_header>* consumed;
+    std::deque<model::record_batch_header>* skipped;
+    // If this is true any call to consume_batch_* or skip_batch_* will
+    // trigger test failure.
+    bool expect_close{false};
+    mutable size_t accept_cnt{0};
+};
+
+SEASTAR_THREAD_TEST_CASE(test_chained_consumers_round_robin) {
+    temporary_dir tmp_dir("chained_consumers");
+    auto tmp_path = tmp_dir.get_path();
+    using namespace storage;
+
+    disk_log_builder b{log_config{
+      tmp_path.string(),
+      1024 * 1024,
+      ss::default_priority_class(),
+      storage::make_sanitized_file_config()}};
+
+    b | start(ntp_config{{"kafka", "panda-topic", 0}, {tmp_path}})
+      | add_segment(0);
+
+    for (int i = 0; i < 20; i++) {
+        b | add_random_batch(i, 1);
+    }
+    auto defer = ss::defer([&b] { b.stop().get(); });
+
+    std::deque<model::record_batch_header> consumed;
+    std::deque<model::record_batch_header> skipped;
+
+    auto c1 = std::make_unique<round_robin_consumer>();
+    c1->side = 0;
+    c1->size_limit = 10000;
+    c1->consumed = &consumed;
+    c1->skipped = &skipped;
+    auto c2 = std::make_unique<round_robin_consumer>();
+    c2->side = 1;
+    c2->size_limit = 10000;
+    c2->consumed = &consumed;
+    c2->skipped = &skipped;
+
+    auto cc = std::make_unique<chained_batch_consumer<2>>(
+      std::move(c1), std::move(c2));
+    auto s = b.get_log_segments().front();
+    auto str = s->offset_data_stream(
+                  s->offsets().get_base_offset(), ss::default_priority_class())
+                 .get();
+
+    auto parser = ss::make_lw_shared<storage::continuous_batch_parser>(
+      std::move(cc), storage::segment_reader_handle(std::move(str)));
+
+    auto close = ss::defer([parser] { parser->close().get(); });
+
+    auto cnt = parser->consume().get();
+    BOOST_REQUIRE(cnt.has_value());
+    BOOST_REQUIRE(cnt.value() > 0);
+    BOOST_REQUIRE(consumed.size() > 0);
+    BOOST_REQUIRE(consumed.size() == skipped.size());
+    BOOST_REQUIRE(consumed == skipped);
+}
+
+SEASTAR_THREAD_TEST_CASE(test_chained_consumers_early_stop) {
+    temporary_dir tmp_dir("chained_consumers");
```

Comment:
Ooh, this is nice! I added get_test_directory() which, for gtests, gets a unique name per test case run. If you don't mind maybe I'll tweak temporary_dir to use it as well (or add a wrapper around it)